hphp/runtime/vm/jit/vasm-reuseimm.cpp

```diff
@@ -0,0 +1,163 @@
+
+// track ldimmq values
+struct ImmState {
+  ImmState() : val{0}, base{} {}
+  ImmState(Immed64 a, Vreg b) : val{a}, base{b} {}
+
+  void reset() { base = Vreg{}; }
+
+  Immed64 val;
+  Vreg base;
+};
+
+struct Env {
+  Vunit& unit;
+  int max_span;
+  std::vector<ImmState> immStateVec;
+};
+
+bool isMultiword(int64_t imm) {
+  switch (arch()) {
+  case Arch::X64:
+  case Arch::PPC64:
+    break;
+  case Arch::ARM:
+    uint64_t val = std::abs(imm);
+    if (val > (1 << 16)) return true;
+    break;
+  }
+  return false;
+}
+
+// candidate around +/- uimm12
+folly::Optional<int> reuseCandidate(Env& env, int64_t p, Vreg& reg) {
+  for (auto const& elem : env.immStateVec) {
+    if (!elem.base.isValid()) continue;
+    int64_t q = elem.val.q();
+    if ((p >= q) && (p < (q + 4095))) {
+      reg = elem.base;
+      return folly::Optional<int>(safe_cast<int>(p - q));
+    }
+    if ((p < q) && (q < (p + 4095))) {
+      reg = elem.base;
+//    return folly::Optional<int>(safe_cast<int>(-safe_cast<int>(q - p)));
+      return folly::Optional<int>(safe_cast<int>(safe_cast<int>(p - q)));
+    }
+  }
+  return folly::none;
+}
+
+template <typename Inst>
+void reuseImmq(Env& env, const Inst& inst, Vlabel b, size_t i) {
+  // leaky bucket
+  env.immStateVec[i % env.max_span] = ImmState{};
+}
+
+template<typename ReuseImm>
+void reuseimm_impl(Vunit& unit, Vlabel b, size_t i, ReuseImm reuse) {
+  vmodify(unit, b, i, [&] (Vout& v) { reuse(v); return 1; });
+}
+
+void reuseImmq(Env& env, const ldimmq& ld, Vlabel b, size_t i) {
+  if (isMultiword(ld.s.q())) {
+    Vreg base;
+    folly::Optional<int> off = reuseCandidate(env, ld.s.q(), base);
+
+    if (off.hasValue()) {
+//    if (off.value() >= 0 ) {
+        reuseimm_impl(env.unit, b, i, [&] (Vout& v) {
+          v << addqi{off.value(), base, ld.d, v.makeReg()};
+        });
+//    } else {
+//      reuseimm_impl(env.unit, b, i, [&] (Vout& v) {
+//        v << subqi{-off.value(), base, ld.d, v.makeReg()};
+//      });
+//    }
+    return;
+    }
+  }
+  env.immStateVec[i % env.max_span] = ImmState{ld.s, ld.d};
+}
+
+void reuseImmq(Env& env, Vlabel b, size_t i) {
+  assertx(i <= env.unit.blocks[b].code.size());
+  auto const& inst = env.unit.blocks[b].code[i];
+
+  if (isBlockEnd(inst) || isCall(inst)) {
+    for (auto &elem : env.immStateVec) elem.reset();
+  } else if (inst.op == Vinstr::ldimmq) {
+    reuseImmq(env, inst.ldimmq_, b, i);
+  } else {
+    reuseImmq(env, inst, b, i);
+  }
+  return;
+}
+
+}
+
+// Opportunistically reuse immediate q values
+void reuseImmq(Vunit& unit) {
+  assertx(check(unit));
+  auto& blocks = unit.blocks;
+
+  if (RuntimeOption::EvalJitLdimmqSpan <= 0) return;
+
+  Env env { unit, RuntimeOption::EvalJitLdimmqSpan };
+  env.immStateVec.reserve(env.max_span);
+
+  auto const labels = sortBlocks(unit);
+
+  for (auto const b : labels) {
+    // required for initial size of vector
+    for (auto i = 0; i < env.max_span; i++) {
+      env.immStateVec.push_back(ImmState{});
```

Comment:
It looks like you're adding `EvalJitLdimmqSpan` entries to this vector for every block that you process, which makes the vector `O(sb)` in memory instead of `O(1)`...

You should do `resize()` above instead of `reserve()`, and then clear the vector at the start of every block like you do above.